azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +100,242 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
+        Policy file is expected to be at conf.get_policy_file_path(). Note that this method should only be called
+        after verifying that the file exists (currently done in __init__).
         """
-        Check whether user has opted into policy enforcement feature.
-        Caller function should check this before performing any operations.
+        with open(conf.get_policy_file_path(), 'r') as f:
+            _PolicyEngine._log_policy_event(
+                "Policy enforcement is enabled. Enforcing policy using policy file found at '{0}'.".format(
+                    conf.get_policy_file_path()))
+            try:
+                # json.load will raise error if file is not in valid json format (including empty file).
+                custom_policy = json.load(f)
+            except ValueError as ex:
+                msg = "policy file does not conform to valid json syntax"
+                raise InvalidPolicyError(msg=msg, inner=ex)
+            except Exception as ex:
+                msg = "unable to load policy file"
+                raise InvalidPolicyError(msg=msg, inner=ex)
+
+            return custom_policy
+
+    def __parse_policy(self, custom_policy):
         """
-        # TODO: Add check for policy file present at /etc/waagent_policy.json.
-        # Policy should only be enabled if conf flag is true AND policy file is present.
-        return conf.get_extension_policy_enabled()
+        Update self._policy with attributes specified in custom_policy:
+            - attributes provided in custom_policy override the default values in self._policy
+            - if an attribute is not provided, use the default value
+            - if an unrecognized attribute is present in custom_policy (not defined in _POLICY_SCHEMA), raise an error
+            - if an attribute does not match the type specified in the schema, raise an error
+
+        If provided, the "extensions" attribute will be converted to a case-folded dict. CRP allows extensions to be any
+        case, so we use case-folded dict to allow for case-insensitive lookup of individual extension policies.
+        """
+        # Validate top level attributes and then parse each section of the policy file.
+        # Individual parsing functions are responsible for validating schema of that section (nested dict).
+        self.__validate_schema(custom_policy, _POLICY_SCHEMA)
+        self.__parse_version(custom_policy)
+        self.__parse_extension_policies(custom_policy)
+
+    def __parse_version(self, policy):
+        """
+        Validate and return "policyVersion" attribute. If not a string in the format "x.y.z", raise InvalidPolicyError.
+        If policy_version is greater than maximum supported version, raise InvalidPolicyError.
+        """
+        version = self.__parse_string(policy, "policyVersion")
+        if version is None:
+            return
+
+        pattern = r'^\d+\.\d+\.\d+$'
```

Comment:
is something like "1" or "1.0" acceptable? or should this be a strict major.minor.patch format?

Reply:
we could use FlexibleVersion

```
>>> FlexibleVersion('1') == FlexibleVersion("1.0")
True
>>> FlexibleVersion('1') == FlexibleVersion("1.0.0")
True
>>> FlexibleVersion('1') == FlexibleVersion("1.0.0.0")
True
>>> FlexibleVersion('1') == FlexibleVersion("1.1")
False

```

Reply:
updated to use FlexibleVersion